src/MongoDB/TopologyDescription.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <Zend/zend_interfaces.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+
+zend_class_entry* php_phongo_topologydescription_ce;
+
+#define PHONGO_TOPOLOGY_UNKNOWN "Unknown"
+#define PHONGO_TOPOLOGY_SINGLE "Single"
+#define PHONGO_TOPOLOGY_SHARDED "Sharded"
+#define PHONGO_TOPOLOGY_REPLICA_SET_NO_PRIMARY "ReplicaSetNoPrimary"
+#define PHONGO_TOPOLOGY_REPLICA_SET_WITH_PRIMARY "ReplicaSetWithPrimary"
+
+/* {{{ proto array MongoDB\Driver\TopologyDescription::getServers()
+   Returns an array of ServerDescription objects for all known servers in the topology */
+static PHP_METHOD(TopologyDescription, getServers)
+{
+	php_phongo_topologydescription_t* intern;
+	mongoc_server_description_t**     sds;
+	size_t                            i, n = 0;
+
+	intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	sds = mongoc_topology_description_get_servers(intern->topology_description, &n);
+
+	array_init_size(return_value, n);
+
+	for (i = 0; i < n; i++) {
+		zval obj;
+
+		php_phongo_server_description_to_zval(&obj, sds[i]);
+		add_next_index_zval(return_value, &obj);
+	}
+
+	mongoc_server_descriptions_destroy_all(sds, n);
+} /* }}} */
+
+/* {{{ proto boolean MongoDB\Driver\TopologyDescription::hasReadableServer()
+    Returns whether the topology has a readable server available */
+static PHP_METHOD(TopologyDescription, hasReadableServer)
+{
+	php_phongo_topologydescription_t* intern          = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+	const mongoc_read_prefs_t*        read_preference = NULL;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	RETVAL_BOOL(mongoc_topology_description_has_readable_server(intern->topology_description, read_preference));
```

Comment:
Should `read_preference` be set to something here or is `NULL` fine (the [docs](http://mongoc.org/libmongoc/current/mongoc_topology_description_has_readable_server.html) say it will just use the default read preferences)? I remember you mentioned looking at `php_phongo_manager_select_server` from the `Manager` class, but that function seems to get the read preference from a provided zval which I don't think I have here.

Reply:
The read preference should be an optional parameter to this method. When we reviewed this earlier in the week I suggested making a note of `Manager::selectServer()`, which also takes a read preference argument for calling [mongoc_client_select_server|http://mongoc.org/libmongoc/1.18.0/mongoc_client_select_server.html].

`Manager::selectServer()`'s argument happens to be required (although it need not be -- see: [PHPC-1937](https://jira.mongodb.org/browse/PHPC-1937)), but there are other examples of parsing optional arguments throughout PHPC. When using `zend_parse_parameters`, everything to the right of a pipe character (`|`) is considered optional. Since this is a new function, I think it'd be a better opportunity to forgo `zend_parse_parameters` and instead use the new PHP 7+ macros, which are documented in [The PHP function structure and API, in C](https://www.phpinternalsbook.com/php7/extensions_design/php_functions.html#the-php-function-structure-and-api-in-c) in the internals book. Andreas also used those when implementing the ServerApi class for Versioned API (see: `ServerApi::__construct`).